pkg/retriever/retrieve.go

```diff
@@ -235,43 +252,133 @@ func runRetrievalCandidate(ctx context.Context, cfg *RetrievalConfig, client Ret
 		timeout = cfg.GetStorageProviderTimeout(candidate.MinerPeer.ID)
 	}
 
-	var stats *types.RetrievalStats
-	var retrievalErr error
-	var done func()
-
 	retrieval.sendEvent(eventpublisher.Started(phaseStartTime, eventpublisher.QueryPhase, candidate))
 
 	// run the query phase
 	onConnected := func() {
 		retrieval.sendEvent(eventpublisher.Connect(phaseStartTime, eventpublisher.QueryPhase, candidate))
 	}
-	queryResponse, queryErr := queryPhase(ctx, cfg, client, timeout, candidate, onConnected)
-	if queryErr != nil {
-		retrieval.sendEvent(eventpublisher.Failure(phaseStartTime, eventpublisher.QueryPhase, candidate, queryErr.Error()))
-	}
+	queryResponse, err := queryPhase(ctx, cfg, client, timeout, candidate, onConnected)
 
+	// we check if both error and response are nil because
+	// it's possible we get both response and err as nil
 	if queryResponse != nil {
 		retrieval.sendEvent(eventpublisher.QueryAsk(phaseStartTime, candidate, *queryResponse))
 		if queryResponse.Status != retrievalmarket.QueryResponseAvailable ||
 			(cfg.IsAcceptableQueryResponse != nil && !cfg.IsAcceptableQueryResponse(queryResponse)) {
-			queryResponse = nil
+
+			// Reduce attempts since we won't be making a retrieval for it
+			attempts := <-retrieval.attemptsChan
```

Comment:
You're essentially using channels like a mutex here, and I'm not sure what that gets you.

Generally, channels should be used to communicate signsls & messages between go-routines to coordinate concurrency, while mutexes are the simplest and fastest way to protect shared state.

Here, you're using channels with buffer 1 in such a way that it's essentially to protect a piece of shared state. I don't get what this gives you over mutexes.

It's not that I'm anti-channel, but the point of channels is to think about how go routines coordinate and signal events to each other, and in doing so, force you to think through the logic of how coordination should happen.

Reply:
> You're essentially using channels like a mutex here, and I'm not sure what that gets you.
Correct, I am using this like a mutex. Honestly, this could be a mutex instead. I would instead have a `attempts int` and an `attemptsMux sync.Mutex` on the retriever state instead of a single attempts channel.

What I liked about using a single buffer channel is the idea that I can pass the data I want to access around in the channel. This guarantees that any access to this data is safe. A routine will request that data, and it now owns that data. It can make updates completely safe. I understand this is no different to properly used mutexes and would not be opposed to using one.

Reply:
Updated to use a mutex